extensions/grafana-loki/PushGrafanaLoki.cpp

```diff
@@ -0,0 +1,249 @@
+  if (auto ssl_context = context.getProperty(PushGrafanaLoki::SSLContextService)) {
+    return std::dynamic_pointer_cast<minifi::controllers::SSLContextService>(context.getControllerService(*ssl_context));
+  }
+  return std::shared_ptr<minifi::controllers::SSLContextService>{};
+}
+
+std::string PushGrafanaLoki::readLogLineFromFlowFile(const std::shared_ptr<core::FlowFile>& flow_file, core::ProcessSession& session) {
+  auto read_buffer_result = session.readBuffer(flow_file);
+  return {reinterpret_cast<const char*>(read_buffer_result.buffer.data()), read_buffer_result.buffer.size()};
+}
+
+void PushGrafanaLoki::setUpStateManager(core::ProcessContext& context) {
+  auto state_manager = context.getStateManager();
+  if (state_manager == nullptr) {
+    throw Exception(PROCESSOR_EXCEPTION, "Failed to get StateManager");
+  }
+  log_batch_.setStateManager(state_manager);
+
+  std::unordered_map<std::string, std::string> state_map;
+  if (state_manager->get(state_map)) {
+    auto it = state_map.find("start_push_time");
+    if (it != state_map.end()) {
+      logger_->log_info("Restored start push time from processor state: {}", it->second);
+      std::chrono::steady_clock::time_point start_push_time{std::chrono::milliseconds{std::stoll(it->second)}};
+      log_batch_.setStartPushTime(start_push_time);
+    }
+  }
+}
+
+std::map<std::string, std::string> PushGrafanaLoki::buildStreamLabelMap(core::ProcessContext& context) {
+  std::map<std::string, std::string> stream_label_map;
+  if (auto stream_labels_str = context.getProperty(StreamLabels)) {
+    auto stream_labels = utils::string::splitAndTrimRemovingEmpty(*stream_labels_str, ",");
+    if (stream_labels.empty()) {
+      throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing or invalid Stream Label Attributes");
+    }
+    for (const auto& label : stream_labels) {
+      auto stream_labels = utils::string::splitAndTrimRemovingEmpty(label, "=");
+      if (stream_labels.size() != 2) {
+        throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing or invalid Stream Label Attributes");
+      }
+      stream_label_map[stream_labels[0]] = stream_labels[1];
+    }
+  } else {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing or invalid Stream Label Attributes");
+  }
+  return stream_label_map;
+}
+
+void PushGrafanaLoki::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
+  setUpStateManager(context);
+  setUpStreamLabels(context);
+
+  if (auto log_line_metadata_attributes = context.getProperty(LogLineMetadataAttributes)) {
+    log_line_metadata_attributes_ = utils::string::splitAndTrimRemovingEmpty(*log_line_metadata_attributes, ",");
+  }
+
+  auto log_line_batch_wait = context.getProperty<core::TimePeriodValue>(LogLineBatchWait);
+  auto log_line_batch_size = context.getProperty<uint64_t>(LogLineBatchSize);
+    if (log_line_batch_size && *log_line_batch_size < 1) {
```

Comment:
strange indentation here

Reply:
Good catch, it seems that some review changes from MINIFICPP-2261 were lost when rebasing this PR. I restored those changes in 970b779746f31f57ea7e95e34008b5cea28ba843